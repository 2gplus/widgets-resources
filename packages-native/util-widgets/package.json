{
    "name": "@native-mobile-resources/util-widgets",
    "version": "0.0.1",
    "private": true,
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "start": "tsc --watch",
        "prebuild": "rm -rf ./dist",
        "build": "tsc",
        "prepare": "npm run build",
        "lint": "tslint --project tsconfig.json --config ../../tslint.json"
    }
}
