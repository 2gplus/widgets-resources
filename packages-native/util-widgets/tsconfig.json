{
    "extends": "@mendix/pluggable-widgets-tools/configs/tsconfig.base",
    "include": ["./src"],
    "compilerOptions": {
        "outDir": "./dist",
        "declaration": true,
        "sourceMap": false
    }
}
