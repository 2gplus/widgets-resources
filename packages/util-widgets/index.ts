export { flattenStyles, Style } from "./src/Styles";
export { only, exclude } from "./src/ObjectFilters";
