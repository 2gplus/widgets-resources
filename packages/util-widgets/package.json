{
    "name": "@native-components/util-widgets",
    "version": "0.0.1",
    "private": true,
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "start": "../../node_modules/.bin/tsc --watch",
        "prebuild": "rm -rf ./dist",
        "build": "../../node_modules/.bin/tsc",
        "prepare": "npm run build",
        "lint": "../../node_modules/.bin/tslint --project tsconfig.json --config ../../tslint.json"
    }
}
