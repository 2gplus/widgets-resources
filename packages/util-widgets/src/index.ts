export { flattenStyles, Style } from "./styles";
export { only, exclude, pluckFromList } from "./ObjectFilters";
