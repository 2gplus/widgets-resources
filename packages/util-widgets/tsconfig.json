{
    "extends": "../../tsconfig.base.json",
    "include": ["../../typings", "./src"],
    "compilerOptions": {
        "outDir": "./dist",
        "declaration": true,
        "sourceMap": false
    }
}
