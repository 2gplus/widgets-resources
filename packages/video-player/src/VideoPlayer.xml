<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.videoplayer.VideoPlayer" pluginWidget="true" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../xsd/widget.xsd">
    <name>VideoPlayer</name>
    <description>My widget description</description>
    <icon/>
    <properties>
        <property key="url" type="attribute">
            <caption>Video URL</caption>
            <category>General</category>
            <description>The URL containing the video. Needs to be MP4</description>
            <attributeTypes>
                <attributeType name="String"/>
            </attributeTypes>
        </property>
    </properties>
</widget>
