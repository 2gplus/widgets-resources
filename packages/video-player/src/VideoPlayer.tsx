import { flattenStyles } from "@native-mobile-resources/util-widgets";
import { Component, createElement } from "react";
import { ActivityIndicator, View } from "react-native";
import Video from "react-native-video";

import { VideoPlayerProps } from "../typings/VideoPlayerProps";
import { defaultVideoStyle } from "./ui/Styles";

export type Props = VideoPlayerProps<undefined>;

interface State {
    loading: boolean;
}

export class VideoPlayer extends Component<Props, State> {
    readonly state = {
        loading: true
    };

    private readonly onLoadStartHandler = this.onLoadStart.bind(this);
    private readonly onLoadHandler = this.onLoad.bind(this);
    private readonly styles = flattenStyles(defaultVideoStyle, this.props.style);

    render(): JSX.Element {
        const uri = this.props.videoUrl && this.props.videoUrl.value;
        return (
            <View style={this.styles.container}>
                {this.state.loading && <ActivityIndicator color={this.styles.indicator.color} size="large" />}
                <Video
                    source={{ uri }}
                    paused={!this.props.autoStart}
                    muted={this.props.muted}
                    repeat={this.props.loop}
                    controls={this.props.showControls}
                    onLoadStart={this.onLoadStartHandler}
                    onLoad={this.onLoadHandler}
                    style={this.state.loading ? { height: 0 } : this.styles.video}
                />
            </View>
        );
    }

    private onLoadStart(): void {
        this.setState({ loading: true });
    }

    private onLoad(): void {
        this.setState({ loading: false });
    }
}
