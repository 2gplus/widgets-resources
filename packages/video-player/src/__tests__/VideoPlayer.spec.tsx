import { dynamicValue } from "@native-components/util-widgets/test";
import { createElement } from "react";
import { render, RenderAPI } from "react-native-testing-library";
import { VideoProperties } from "react-native-video";

import { Props, VideoPlayer } from "../VideoPlayer";
import Video from "./__mocks__/Video";

jest.mock("react-native-video", () => require.requireActual("./__mocks__/Video"));

describe("VideoPlayer", () => {
    let defaultProps: Props;

    beforeEach(() => {
        defaultProps = {
            style: [],
            videoUrl: dynamicValue("https://mendix.com/video.mp4"),
            autoStart: false,
            muted: false,
            loop: false,
            showControls: true
        };
    });

    it("renders a loading indicator", () => {
        const component = render(<VideoPlayer {...defaultProps} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("passes the right props to the video player", () => {
        const component = render(<VideoPlayer {...defaultProps} />);
        const props = videoInstanceProps(component);

        expect(props.source).toEqual({ uri: "https://mendix.com/video.mp4" });
        expect(props.paused).toBe(true);
        expect(props.muted).toBe(false);
        expect(props.repeat).toBe(false);
        expect(props.controls).toBe(true);
        expect(props.style).toEqual({ height: 0 });
    });

    it("hides the loading indicator after load", () => {
        const component = render(<VideoPlayer {...defaultProps} />);

        videoInstanceProps(component).onLoad!({} as any);

        expect(component.toJSON()).toMatchSnapshot();
        expect(videoInstanceProps(component).style).toEqual({ width: "100%", height: "100%" });
    });

    it("shows the loading indicator if the source changes", () => {
        const component = render(<VideoPlayer {...defaultProps} />);

        videoInstanceProps(component).onLoad!({} as any);
        videoInstanceProps(component).onLoadStart!();

        expect(component.toJSON()).toMatchSnapshot();
        expect(videoInstanceProps(component).style).toEqual({ height: 0 });
    });
});

function videoInstanceProps(component: RenderAPI): VideoProperties {
    const RNCameraInstance = component.getByType(Video);
    return RNCameraInstance.props as VideoProperties;
}
