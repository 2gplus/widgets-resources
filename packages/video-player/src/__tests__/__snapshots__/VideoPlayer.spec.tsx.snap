// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer hides the loading indicator after load 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "flex": 1,
      "justifyContent": "center",
    }
  }
/>
`;

exports[`VideoPlayer renders a loading indicator 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="white"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`VideoPlayer shows the loading indicator if the source changes 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="white"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
